docs/readmes/proposals/p017_mme_migrate_to_c++.md

```diff
@@ -0,0 +1,133 @@
+
+
+### State Serialization
+The serialization performance of protobuf is insufficient, so a serialization library that allows better performance must be selected.
+
+The bad performance is due to the global serialization process (translation, memory allocation, copy).
+
+The replacement libraries considered are "cap'n proto" and "Flat buffers”, they support C/C++ standart API, binary output format and do not need memory copy (zero copy).
+
+
+The states that have to be serialized are (To be detailed):
+- S1AP, S1AP_IMSI_MAP, S1AP_UE
+- MME_UE, MME_NAS
+- SGW, SGW_UE, SPGW, SPGW_UE
+- S11/GTPV2-C
+- NGAP, NGAP_IMSI_MAP, NGAP_UE
+
+#### S1AP, MME_APP, NAS migration
+
+ S1AP seems to have no state dependency with other tasks.
+ 
+ Should be an opportunity to start doing a first migration with S1AP.
+
+ MME_APP and NAS tasks have contexts in common.
+ 
+
+////### Migrate some MM and NAS IEs
+////For the ease of manipulating IEs, migrate a selected list of IEs (operator ==, copy, conversion, etc)
+
+
+### 3GPP protocols (OAI)
+This work is necessary only for the deployment of Magma MME with CUPS architecture.
+
+S6a Interface implemented in C should be migrated to C++.
+A C++ port of S6a interface is already available in OAI HSS repo (https://github.com/OPENAIRINTERFACE/openair-hss) could serve as a basis.
+
+S11 Interface and its GTPV2-C protocol implemented in C should be migrated to C++.
+A C++ port of S11 and GTPV2-C interface done in OAI SPGW-C repo (https://github.com/OPENAIRINTERFACE/openair-spgwc) could serve as a basis.
+
+
+## Rationale
+
+### Get/Set state attributes
+
+For a progressive migration, and a decoupling of MME core code from the serialization library, macros, inline, helpers functions/C++ patterns may be used instead of directly using serialization library API in MME core code.
+
+Some MM and NAS IEs, for the ease of their manipulation (comparison operators, etc) could be migrated with wrapper classes.
+
+[A discussion of alternate approaches and the trade offs, advantages, and
+disadvantages of the specified approach.]
+
+## Compatibility
+
+Endianness: Whatever is the HW platform, endianness is fixed.
+
+Backward compatibility of the new serialization library with protobuf is not envisioned.
+
+## Observability and Debug
+
+TODO
+[A description, of how issues with this design would be observed and debugged
+in various stages from development through production.]
+
+## Implementation
+
+  Description of steps in the C++ migration:
+ 
+* Serialization library selection
+    * Build 2 realistics Prototypes with MME_APP, NAS protos (cap’n proto, Flat buffers)
+        * Migrate hashtables (find functions)
+    * Get measurements and evaluate performances, maintainability.
+    * Select library
+* MME_APP/NAS Serialization migration
+    * Write proto
+    * Write redis client libraries
+    * Migrate hashtables (find functions, evaluate performance)
```

Comment:
@lionelgo I would move this as a separate effort. Maybe instead of hashtable migration being part of the serialization efforts, I think it should be in its own contained task.

Reply:
@ardzoht I change accordingly to your suggestion, It will be clearer to me when I will get more familiar with the serialization library.

Reply:
@ardzoht I made a commit, if I missed your point, please change task list according to your thoughts